### ISSUE TITLE
Replace  pragma "no doc" with something better

### ISSUE THREAD

**Issue by ronawho** [chapel-lang/chapel#13115]
Having to use `pragma "no doc"` feels like a kludge for something that we (and probably users) use pretty often.

My preference would be to not generate docs for procedures that don't have a chpldoc comment attached to them. This seems like the common case and if a user wants to expose the signature without adding text they could just add an empty comment:

```chpl
// don't generate chpldoc
proc foo(arg: type): retType { }

// generate chpldoc
/* */
proc foo(arg: type2): retType { }
```

There could also be cases where users want to document for internal usage, but don't want to expose it publicly. For that we could have a special sentinel in the chpldoc comment. Maybe something like:

```chpl
// don't generate chpldoc
/* Developer documentation
   :nodoc
 */
proc foo(arg: type): retType { }
```

(another option would be to take the javadoc approach and use `/**` to denote procedures that should be publicly documented)

**Comment by bradcray** [comment]
Just to make sure anyone coming across this is aware, marking such cases as `private` will also squash their `chpldoc`umentation.

**Comment by lydia-duncan** [comment]
I would like to allow users to write things like:
```chapel
// generate chpldoc
/* */
proc foo(arg: type): retType { }
proc foo(arg: type2): retType { }
```
and have both of the functions share the same documentation without having to copy the documentation comment simply to preserve the second function in the generated documentation

**Comment by lydia-duncan** [comment]
We already do this with some functions in our documentation today.

**Comment by ronawho** [comment]
I'm not sure I follow. Today:

```chpl
/* Documentation */
proc foo(arg: type1): retType { }
proc foo(arg: type2): retType { }
```

doesn't really "share" documentation. One proc is documented and the second one follows the documentation of the first:

![Screen Shot 2019-05-29 at 2 30 38 PM](https://user-images.githubusercontent.com/1588337/58592908-71b7a000-821e-11e9-979f-af119419b7a0.png)

If they actually shared documentation I would imagine you'd get both functions listed, and then the documentation for them.

But still I think it's much more common to want to avoid generating documentation for a function without a chpldoc comment, than it is to want to document a function without a chpldoc comment. Having to add `/* */` before some overloaded functions seems much easier than having to "no doc" as many functions as we currently do

**Comment by lydia-duncan** [comment]
I probably did reverse that order.

**Comment by lydia-duncan** [comment]
I don't think it is reasonable to treat the internal or even the standard and package modules as a typical use case for documentation.  Far more of the symbols there are meant to be undocumented due to trying to eat our own dogfood (not that I am arguing against having that part of the implementation in module code by any means).

Additionally, if a user finds themselves wanting to `no doc` large swathes of their code, maybe that speaks to organizing that code under a helper module that itself gets `no doc` applied to it.  I would much prefer considering different code organizations to flipping a switch and having to go through all of our documentation to ensure we didn't accidentally lose something by doing so.

**Comment by ronawho** [comment]
I think standard and package modules are fair game, though I agree internal modules are different.

In my experience if I'm going to expose a function to a user I will put a chpldoc comment on it, or it's an overload for a different function I'm going to overload. I'm having a hard time coming up with a lot of use-cases where I want to expose a function signature without attaching documentation to it (but I don't write that much chpl code)

**Comment by lydia-duncan** [comment]
Okay, here's my main philosophy:

Choosing to `"no doc"` a symbol rather than make it `private` is something that should happen rarely.  It pollutes the client's namespace.  If it is necessary for functionality, they are probably doing some weird code organization and should either a) consider whether they actually do want to expose the functionality to the user, or b) reorganize their code such that they can make the symbol private.  By having `"no doc"` be something the user has to do actively (as opposed to something that happens for them by default), it makes them more likely to think about the consequences of doing so (especially if it becomes painful because they are doing it so often).

**Comment by lydia-duncan** [comment]
Now, maybe symbols should be `private` by default.  Maybe `use`s of modules in A shouldn't impact scopes that `use A`.  I think that is independent of documenting or not documenting a symbol - I think if a symbol is `public`, it should be documented.

**Comment by ronawho** [comment]
There's a lot of things we can't make private today, but I guess we plan to support that eventually: https://github.com/chapel-lang/chapel/issues/6067

It seems weird to me to use documentation as a forcing function to get users thinking about namespace visibility. I would guess the more likely outcome is that we will generate documentation for things a user didn't want to be documented and they won't find out for a while (we have done this ourselves because we don't look at the generated documentation for every change we make to a module -- https://github.com/chapel-lang/chapel/pull/11110). There's also valid use-cases for something that can't be private but isn't meant for users like shared helper routines meant for internal usage. 

But I don't think I'm going to convince you that documentation should be opt-in instead of opt-out. Assuming the default stays opt-out -- I still want something more user-facing than a pragma. Even something like `/* :nodoc */` would be nicer to me.

**Comment by lydia-duncan** [comment]
> I still want something more user-facing than a pragma.

I 100% agree.  I think the closest we've come to a suggestion that is better than the current status quo is `/* :nodoc: */`, but that either:
A) requires generating the symbol and then having Sphinx remove it for us when we create the html, which fails to help the files it depends on or
B) means performing analysis on the contents of a comment in the chpl compiler stage before we generate the files for Sphinx, which is not something the compiler has done before.

Maybe I should just get over my distaste for B, though

**Comment by bradcray** [comment]
> There's also valid use-cases for something that can't be private but isn't meant for users like shared helper routines meant for internal usage.

I think once we have `private use` implemented, this shouldn't be an issue anymore.  You'd just only privately `use` the helper routine's module.  Does that seem right?

**Comment by lydia-duncan** [comment]
Almost.  I think what you mean is that the helper module itself should be `private`, and then it won't get documented by default?

**Comment by mppf** [comment]
I like `/* :nodoc: */` for what it's worth.

**Comment by bradcray** [comment]
> Almost. I think what you mean is that the helper module itself should be private, and then it won't get documented by default?

I suppose it depends on what you mean by "should be private".  I was just imagining that we wouldn't invoke chpldoc on things that we didn't want documented.

**Comment by ronawho** [comment]
I originally came into this thinking "Why do I have to 'no doc' something I haven't put documentation on?". But it sounds like the right attitude should be "Why do I have something public that's not documented?" (it's accessible to users, so we should expose that fact in the docs)

Most of my experience with writing non-trivial amounts of user documentation was with javadoc where you only generate docs for methods with javadoc comments on them. As a quick survey -- It looks like rust and go also follow the opt-in approach, but Python is opt-out (things with leading underscores do not generate documentation though.)

Conceding the "I want docs to be opt-in" part, I'd be in favor of something like `/* :nodoc: */`.

**Comment by ben-albrecht** [comment]
As an observation, `:nodoc:` as the syntax doesn't really make sense in the context of restructured text. In rst, the [`` :role:`content` ``](http://docutils.sourceforge.net/docs/ref/rst/roles.html) syntax wouldn't impact any content outside of the backticked content region.

Something like a directive would make more sense in rst:

```chpl
/*
.. nodoc::
  this is a private function
*/
proc secret() { }
```

However, I would not advocate for this syntax, since it's fragile w.r.t. indentation.  Also, it wouldn't make a difference implementation-wise because the Chapel compiler would still need to parse the rst to prevent  the generation of the function documentation anyway.

**Comment by mppf** [comment]
What if it were literally
``` chapel
/**/ proc secret() { }
```

? 

Or what about
``` chapel
/* nodoc */ proc secret() { }
```

?

**Comment by lydia-duncan** [comment]
I'd prefer Michael's second choice.  Or maybe:
```chapel
/* chpldoc: no doc */ proc secret() { }
```
Trying to strike a balance between unique enough that users won't accidentally trigger it, and short enough that it isn't a huge burden to write

**Comment by lydia-duncan** [comment]
I think this can be closed with the addition of the "nodoc" pragma, right Elliot?

I should open a separate issue to track my desire for sharing documentation across multiple similar functions without having to copy+paste it, but I should have done that anyways

**Comment by ronawho** [comment]
If you mean the `@chpldoc.nodoc` syntax, yeah I agree

**Comment by lydia-duncan** [comment]
Oops, yeah, meant to say attribute but obviously was thinking about pragmas.
